### ISSUE TITLE
Shop: Product filtering by taxon broken

### ISSUE THREAD

**Issue by aleho** [Sylius/Sylius#12792]
**Sylius version affected**: 1.9+ (probably others too)

**Description**
https://github.com/Sylius/Sylius/blob/710fedab3e556522466131ec819cbe0c81be5e2f/src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php#L64-L86
In `ProductRepository::createShopListQueryBuilder()` there's an `addSelect('productTaxon')` which breaks retrieval of results.

When a product is in multiple taxons it will be returned multiple times by the query (as soon as a join to taxons is done).

This is why you used a `distinct()`.

Distinct though will not filter by IDs, it will compare all results (very slow, by the way) and as soon as the productTaxon is selected in the result rows, distinct will not filter it out.

So there will be n results for each page, but the Doctrine hydrator will reduce these to unique products returned. A page with limit 12 e.g. had only 7 results here.

**Steps to reproduce**  
Open a taxon index page with `$includeAllDescendants = true`.

**Possible Solution**  
Either don't use `distinct`, use `groupBy('o.id')` or
don't use `addSelect('productTaxon')` (or probably both).

**Comment by lchrusciel** [comment]
For anyone also having issues with this problem. Another possibility to fix it is to turn on `OutputWalkers` and `fetchCollectionJoins` on DoctrineAdapter for Pagerfanta. In order to do it you should:

1. Create a \App\Grid\DataSource class with this body:
```php
<?php

declare(strict_types=1);

namespace App\Grid;

use Doctrine\ORM\QueryBuilder;
use Pagerfanta\Adapter\DoctrineORMAdapter;
use Pagerfanta\Pagerfanta;
use Sylius\Bundle\GridBundle\Doctrine\ORM\ExpressionBuilder;
use Sylius\Component\Grid\Data\DataSourceInterface;
use Sylius\Component\Grid\Data\ExpressionBuilderInterface;
use Sylius\Component\Grid\Parameters;

final class DataSource implements DataSourceInterface
{
    /** @var QueryBuilder */
    private $queryBuilder;

    /** @var ExpressionBuilderInterface */
    private $expressionBuilder;

    public function __construct(QueryBuilder $queryBuilder)
    {
        $this->queryBuilder = $queryBuilder;
        $this->expressionBuilder = new ExpressionBuilder($queryBuilder);
    }

    public function restrict($expression, string $condition = DataSourceInterface::CONDITION_AND): void
    {
        switch ($condition) {
            case DataSourceInterface::CONDITION_AND:
                $this->queryBuilder->andWhere($expression);

                break;
            case DataSourceInterface::CONDITION_OR:
                $this->queryBuilder->orWhere($expression);

                break;
        }
    }

    public function getExpressionBuilder(): ExpressionBuilderInterface
    {
        return $this->expressionBuilder;
    }

    public function getData(Parameters $parameters)
    {
        // Use output walkers option in DoctrineORMAdapter should be false as it affects performance greatly. (see #3775)
        $paginator = new Pagerfanta(new DoctrineORMAdapter($this->queryBuilder, false, false));
        $paginator->setNormalizeOutOfRangePages(true);
        $paginator->setCurrentPage($parameters->get('page', 1));

        return $paginator;
    }
}
```
2. Create a \App\Grid\Driver class with:
```php
<?php

declare(strict_types=1);

namespace App\Grid;

use Doctrine\ORM\EntityRepository;
use Doctrine\Persistence\ManagerRegistry;
use Doctrine\Persistence\ObjectManager;
use Sylius\Component\Grid\Data\DataSourceInterface;
use Sylius\Component\Grid\Data\DriverInterface;
use Sylius\Component\Grid\Parameters;

final class Driver implements DriverInterface
{
    public const NAME = 'doctrine/orm';

    /** @var ManagerRegistry */
    private $managerRegistry;

    public function __construct(ManagerRegistry $managerRegistry)
    {
        $this->managerRegistry = $managerRegistry;
    }

    public function getDataSource(array $configuration, Parameters $parameters): DataSourceInterface
    {
        if (!array_key_exists('class', $configuration)) {
            throw new \InvalidArgumentException('"class" must be configured.');
        }

        /** @var ObjectManager $manager */
        $manager = $this->managerRegistry->getManagerForClass($configuration['class']);

        /** @var EntityRepository $repository */
        $repository = $manager->getRepository($configuration['class']);

        if (!isset($configuration['repository']['method'])) {
            return new DataSource($repository->createQueryBuilder('o'));
        }

        $arguments = isset($configuration['repository']['arguments']) ? array_values($configuration['repository']['arguments']) : [];
        $method = $configuration['repository']['method'];
        if (is_array($method) && 2 === count($method)) {
            $queryBuilder = $method[0];
            $method = $method[1];

            return new DataSource($queryBuilder->$method(...$arguments));
        }

        return new DataSource($repository->$method(...$arguments));
    }
}
```
3. Override service definition(e.g. in config/services.yaml):
```yaml
    sylius.grid_driver.doctrine.orm:
        class: 'App\Grid\Driver'
        arguments:
            - '@doctrine'
        tags:
            -
                name: 'sylius.grid_driver'
                alias: 'doctrine/orm'
```

Keep in mind, that this configuration may have a significant performance impact. However, according to my research, it was not noticeable up to 500k simple products being only in one taxon.

Ref.
* https://github.com/Sylius/Sylius/issues/3775
* https://github.com/Sylius/Sylius/issues/3836
* https://github.com/Sylius/Sylius/pull/6674
* https://github.com/doctrine/orm/issues/7211
* https://github.com/vvasiloi/doctrine-pagination
* https://github.com/Sylius/SyliusGridBundle/pull/68

**Comment by aleho** [comment]
As a small comment on my proposed "fixes": Grouping by only one of the output columns is non-standard SQL may be supported by your database (MariaDB) or not (MySQL). For us it works.

**Comment by lchrusciel** [comment]
Should be already fixed with SyliusGridBundle v1.10
